### PR TITLE
Add Compose instance for Map

### LINKED ISSUE #2399: Add a `Compose` instance for `Map`
Should be relatively easy to define and lawful as well :)

### PR DESCRIPTION
**Pull request by denisrosca**
Closes #2399

### CODE REVIEW

File: tests/src/test/scala/cats/tests/MapSuite.scala [hunk 1]
```diff
@@ -15,6 +15,8 @@ class MapSuite extends CatsSuite {
   checkAll("Map[Int, Int] with Option", UnorderedTraverseTests[Map[Int, ?]].unorderedTraverse[Int, Int, Int, Option, Option])
   checkAll("UnorderedTraverse[Map[Int, ?]]", SerializableTests.serializable(UnorderedTraverse[Map[Int, ?]]))
 
+  checkAll("Compose[Map]", ComposeTests[Map].compose[Int, Long, String, Double])
```
Comment (kailuowang) [inline tests/src/test/scala/cats/tests/MapSuite.scala]: nit, can we add the `SerializableTests` as the other instances here?

File: core/src/main/scala/cats/instances/map.scala [hunk 2]
```diff
@@ -79,4 +80,23 @@ trait MapInstances extends cats.kernel.instances.MapInstances {
 
     }
   // scalastyle:on method.length
+
+}
+
+trait MapInstancesBinCompat0 {
+
+  implicit def catsStdComposeForMap: Compose[Map] = new Compose[Map] {
```
Comment (ceedubs) [inline core/src/main/scala/cats/instances/map.scala]: Should this be a `val` instead of a `def` so the same instance can be reused repeatedly?

**Comment by denisrosca** [comment]
@ceedubs Would a doctest like this be enough?

    /**
      * Compose two maps `g` and `f` by using the values in `f` as keys for `g`.
      * {{{
      * scala> import cats.arrow.Compose
      * scala> import cats.implicits._
      * scala> val first = Map(1 -> "a", 2 -> "b", 3 -> "c", 4 -> "a")
      * scala> val second = Map("a" -> true, "b" -> false, "d" -> true)
      * scala> Compose[Map].compose(second, first)
      * res0: Map[Int, Boolean] = Map(1 -> true, 2 -> false, 4 -> true)
      * }}}
      */
      def compose[A, B, C](f: Map[B, C], g: Map[A, B]): Map[A, C]

**Comment by ceedubs** [comment]
@denisrosca that example looks great!

**Pull request merged**